crates/parcel_core/src/plugin/transformer_plugin.rs

```diff
@@ -26,103 +25,70 @@ pub struct ResolveOptions {
-  ) -> Result<Vec<Asset>, anyhow::Error>;
-
-  // Perform processing after the transformation
-  fn post_process(
-    &mut self,
-    config: &PluginConfig,
-    assets: Vec<&Asset>,
-  ) -> Result<Vec<Asset>, anyhow::Error>;
-
-  /// Stringify the AST
-  ///
-  /// This function is called when the next transformer AST cannot be reused, or this is the last
-  /// transformer in a pipeline.
-  ///
-  fn generate(&self, asset: Asset, ast: AST) -> Result<GenerateOutput, anyhow::Error>;
+/// * File-paths that have just been discovered
+/// * Outputs of previous transformation steps, which are in-place modified
+/// * These two scenarios are distinguished
+pub enum TransformationInput {
+  FilePath(PathBuf),
+  Asset(Asset),
 }
 
-#[cfg(test)]
-mod tests {
-  use super::*;
-
-  #[derive(Debug)]
-  struct TestTransformerPlugin {}
-
-  impl TransformerPlugin for TestTransformerPlugin {
-    fn can_reuse_ast(&self, _ast: AST) -> bool {
-      todo!()
-    }
-
-    fn parse(
-      &mut self,
-      _config: &PluginConfig,
-      _asset: &Asset,
-      _resolve: &Resolve,
-    ) -> Result<AST, anyhow::Error> {
-      todo!()
+impl TransformationInput {
+  pub fn file_path(&self) -> &Path {
+    match self {
+      TransformationInput::FilePath(path) => path,
+      TransformationInput::Asset(asset) => asset.file_path(),
     }
+  }
 
-    fn transform(
-      &mut self,
-      _config: &PluginConfig,
-      _asset: &mut Asset,
-      _resolve: &Resolve,
-    ) -> Result<Vec<Asset>, anyhow::Error> {
-      todo!()
+  pub fn read_source_code(&self, fs: FileSystemRef) -> anyhow::Result<Rc<SourceCode>> {
+    match self {
+      TransformationInput::FilePath(file_path) => {
+        let code = fs.read_to_string(&file_path)?;
+        let code = SourceCode::from(code);
+        Ok(Rc::new(code))
+      }
+      TransformationInput::Asset(asset) => Ok(asset.source_code.clone()),
     }
+  }
+}
 
-    fn post_process(
-      &mut self,
-      _config: &PluginConfig,
-      _assets: Vec<&Asset>,
-    ) -> Result<Vec<Asset>, anyhow::Error> {
-      todo!()
-    }
+/// Context parameters for the transformer, other than the input.
+pub struct RunTransformContext {
+  file_system: FileSystemRef,
+}
 
-    fn generate(&self, _asset: Asset, _ast: AST) -> Result<GenerateOutput, anyhow::Error> {
-      todo!()
-    }
+impl RunTransformContext {
+  pub fn new(file_system: FileSystemRef) -> Self {
+    Self { file_system }
   }
 
-  #[test]
-  fn can_be_defined_in_dyn_vec() {
-    let mut transformers = Vec::<Box<dyn TransformerPlugin>>::new();
+  pub fn file_system(&self) -> FileSystemRef {
+    self.file_system.clone()
+  }
+}
 
-    transformers.push(Box::new(TestTransformerPlugin {}));
+#[derive(Debug)]
+pub struct TransformResult {
+  pub asset: Asset,
+  pub dependencies: Vec<Dependency>,
+  /// The transformer signals through this field that its result should be invalidated
+  /// if these paths change.
+  pub invalidate_on_file_change: Vec<PathBuf>,
```

Comment:
I personally find the existing invalidations API tedious to work with and don't think it should be modelled like this long-term. There is a common `Invalidation` type that the request tracker works off, which plugins could return and it would enable all types of invalidations. Whether we want to scope them down per plugin is a different question though. Anyways, just food for thought.

Reply:
But that's just an empty enum?

I see two ways to model this:

1. Dynamic - we have a method on "Run...Context" that pushes this out. This is the approach on current `RequestAPI`

* Advantage, incremental, works for error types

```
context.invalidate_on_file_change(...)
```
2. Static - we have a return value ; this is what is written here, and is the same as `enum Invalidation`

* Advantage, clear typing, less exit points

---

I'm not sure which is better, but think this is quite easy to change as things take form.

Reply:
I expanded this on the `core-rs2` branch like this: https://github.com/parcel-bundler/parcel/blob/103a063b8f526690cf9eb514b75c9da1ff96960a/crates/core/src/request_tracker.rs#L113-L122

I like the idea of just returning them as part of plugin results rather than mutating an input.